accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+protocol by hand if necessary.
+
+## Debug Adapter Components
+
+As we would like the Wasmtime debugger to be able to understand the source
+language of any Wasm that it’s executing, we’re presented with a problem: how
+can we foresee all possible kinds of debuggee programs, compiled or interpreted,
+and provide a unified debugging interface to them? Instead of solving this
+impossible problem, we propose using the component model to design an interface
+for a debug adapter. A debug adapter will import a private interface that gives
+it low-level access to the underlying Wasm debuggee, and will export a
+higher-level interface that the debugger will interact with and query. This
+allows us to uphold our principle of supporting debugging anything compiled to
+Wasm; we just need a debug adapter component to bridge the gap between the
+high-level source language and low-level Wasm semantics.
+
+We will implement two adapters:
+
+1. **The Null Debug Adapter:** A debug adapter that forwards all queries
+   directly to the underlying Wasm being executed. This provides Wasm-level
+   debugging to the developer.
+
+2. **The DWARF Debug Adapter:** A debug adapter component that translates
+   queries from the debugger through the source program’s DWARF debug info
+   custom sections. This enables the debugging of programs compiled to Wasm,
+   such as Rust, C, C++, etc… 
+
+Interpreters that are compiled to Wasm can, in the future, provide their own
+debug adapter components that translate debugging queries through their runtime
+data structures to inspect debuggee programs being interpreted within
+themselves.
+
+Even further into the future, we can collaborate with the WebAssembly Community
+Group to standardize debug adapter components (or something evolved from them)
+in the ``WebAssembly/tool-conventions`` and `WebAssembly/component-model`
+repositories.
+
+While the exact WIT interface of and world for debug adapter components is out
+of scope for this RFC, and will require experimentation and prototyping during
+implementation, we expect that it will closely mirror the kinds of queries
+supported by the Debug Adapter Protocol.
+
+## Live Debugging
+
+Live debugging support comprises several features, each of which comes with
+design and implementation considerations. These features are: setting
+breakpoints, single stepping, watch points, step into/out of, and integration
+with our existing baseline compiler Winch.
+
+### Breakpoints and Single-stepping
+
+To implement breakpoints and single-stepping, we need the ability to pause
+program execution and wait for user input prior to executing a specific line of
+code. The way this is handled in native debuggers is as follows:
+
+* When a user sets a breakpoint, the instruction corresponding to that line of
+  code is replaced with an `int3` instruction.
+
+* When executed, the `int3` instruction raises a `SIGTRAP`, halting execution
```

Comment:
The same thing here when it comes to `SIGTRAP`: things work differently on Windows, for instance, where an `EXCEPTION_BREAKPOINT` is generated instead.  Maybe use some less-platform-specific language?